### ISSUE TITLE
Imported propTypes triggers false positives

### ISSUE THREAD

**Issue by monsieurnebo** [yannickcr/eslint-plugin-react#2357]
Hello,

## Context
Package | Version
--- | ---
eslint | ~6.0.1
eslint-plugin-react | ~7.14.2

## The situation
After updating ESLint, I'm encountering false positives when importing propTypes (that are used in multiple places, for instance).

Let's take a simple example:
```js
// sharedPropTypes.js

import PropTypes from "prop-types";

export default {
  steps : PropTypes.array,
};
```

```jsx
// MyComponent.jsx

import React from "react";
import sharedPropTypes from "./sharedPropTypes";

const propTypes = {
  ...sharedPropTypes,
};

function MyComponent(props) {
  // Using steps.map somewhere
}
```

This scenario is triggering the following lint errors:
`error  'steps.map' is missing in props validation   react/prop-types`

Which is new.

Please note that declaring propTypes in the component file is working fine:
```jsx
// MyComponent.jsx

import React from "react";
import PropTypes from "prop-types";
import sharedPropTypes from "./sharedPropTypes";

const propTypes = {
  steps : PropTypes.array
};

function MyComponent(props) {
  // Using steps.map somewhere
}
```

## A lead
After some searches, it seems that ESLint is more sensitive since https://github.com/yannickcr/eslint-plugin-react/pull/2301 . 

1. Could it be the source?
2. How should I manage this situation in a clean way?

## Related issues
https://github.com/yannickcr/eslint-plugin-react/issues/2352
https://github.com/yannickcr/eslint-plugin-react/issues/2343

**Comment by Borovensky** [comment]
Have the same issue.
`eslint: 5.16.0`
`eslint-plugin-react: 7.14.3`

**Comment by Borovensky** [comment]
@monsieurnebo Hi!
I play a bit around it and this is what I have: 
In my case I have this issue when I declaring propTypes in the component file:
``` javascript
// MyComponent.jsx
import React from "react";
import PropTypes from "prop-types";

const propTypes = {
  steps: PropTypes.array
};

function MyComponent(props) {
  // Using steps.map somewhere
}
```
but if I change the default import of propTypes to a single export from the "prop-types" module the error is gone.
``` javascript
// MyComponent.jsx
import React from "react";
import { array } from "prop-types";

const propTypes = {
  steps: array
};

function MyComponent(props) {
  // Using steps.map somewhere
}
```
Hope this will help you also.

**Comment by ljharb** [comment]
You should never have to destructure off of PropTypes to have things work; that’s a bug.

The OP example with sharedPropTypes just won’t work with static analysis; I’d suggest not sharing the entire propTypes object and instead sharing only individual shapes or propTypes.

**Comment by monsieurnebo** [comment]
@ljharb Thanks for the answer.

Why is this error showing now, and not before?

**Comment by monsieurnebo** [comment]
That's really strange, because I have the same situation in another project, without the lint error...

There must be a difference, that could explain the error... But I'm struggling to find it.

**Comment by ljharb** [comment]
Where did you assign propTypes to MyComponent.propTypes in the SFC version?

**Comment by monsieurnebo** [comment]
Sorry, I missed the props assignation in the pseudocode. It looks like this:

```jsx
import React from "react";
import sharedPropTypes from "./sharedPropTypes";

const propTypes = {
  ...sharedPropTypes,
};

function MyComponent(props) {
  // Using steps.map somewhere
}

MyComponent.propTypes = propTypes;
```

**Comment by ljharb** [comment]
what happens if you do `MyComponent.propTypes = { ...sharedPropTypes }`?

**Comment by monsieurnebo** [comment]
Same error.

**Comment by ljharb** [comment]
In that case it really does seem like we're doing something different for classes and SFCs, so that's a bug.

**Comment by monsieurnebo** [comment]
Good to know!

**Comment by jstheoriginal** [comment]
@monsieurnebo can you confirm this is only happening inside of a nested function you’ve made that contains JSX and gets called inside of the return of the component? We had this happening in this scenario; but not if the prop is accessed directly in the return of the component.

Eslint seems to think the nested function is a SFC component itself so it thinks it’s missing props in the parameter of the nested function.

**Comment by monsieurnebo** [comment]
@jseminck I'm not sure to fully understand your scenario 🤔

**Comment by jstheoriginal** [comment]
> @jseminck I'm not sure to fully understand your scenario 🤔
> Could you provide a simple example of what you're talking about?

Here's an example:

```
interface Props {
  data: {
    text: string,
    icon: Image,
}

const Button = ({ data }: Props) => {
  ...

  const renderLabel = () => <Text>{data.text}</Text>;

  return (
    <View>
      {renderLabel()}
    </View>
  );
};
```

This ESLint rule should complain that data.text is missing in props validation. The reason seems to be because ESLint thinks that renderLabel const is a different component of its own and needs to have props of its own. The error goes away if you pass the props in to the parameters of renderLabel.

**Comment by monsieurnebo** [comment]
Thanks for the clarification, I understand it better. 
Yes, I confirm that's happening in a nested function such as:

```jsx
import React from "react";
import sharedPropTypes from "./sharedPropTypes";

const propTypes = {
  ...sharedPropTypes,
};

function MyComponent(props) {
  const renderFoo = () => {
    <div>
      {props.steps.map(item => ...)}
    </div>
  }

  return (
    <div>
      {renderFoo()}
    </div>
  );
}

MyComponent.propTypes = propTypes;
```

And the error goes away by passing the props as you said:

```jsx
import React from "react";
import sharedPropTypes from "./sharedPropTypes";

const propTypes = {
  ...sharedPropTypes,
};

function MyComponent(props) {
  const renderFoo = (passedProps) => {
    <div>
      {passedProps.steps.map(item => ...)}
    </div>
  }

  return (
    <div>
      {renderFoo(props.steps)}
    </div>
  );
}

MyComponent.propTypes = propTypes;
```

Passing props to nested functions can be a turnaround for now, but it would be nice that ESLint understands it by itself (if that's something doable 🤔 ).

**Comment by saunders1989** [comment]
I was having the same issue on version `7.16.0` a work around for me was to downgrade to `7.13.0` hopefully this will be fixed

**Comment by jzabala** [comment]
@ljharb I think we can close this one. As you said in https://github.com/yannickcr/eslint-plugin-react/issues/2357#issuecomment-514675480 importing propTypes won't work with static analysis and using a variable, defined in the same file, was fixed by #2704
